nx/lib/nx/defn/compiler.ex

```diff
@@ -120,12 +123,30 @@ defmodule Nx.Defn.Compiler do
   def __compile__(fun, params, opts) do
     {compiler, runtime_fun, opts} = prepare_options(fun, opts)
     compiler.__compile__(fun, params, runtime_fun, opts)
+  rescue
+    e in [UndefinedFunctionError] ->
+      raise_missing_callback(e, :__compile__, 4, __STACKTRACE__)
   end
 
   @doc false
   def __jit__(fun, params, args_list, opts) do
     {compiler, runtime_fun, opts} = prepare_options(fun, opts)
     compiler.__jit__(fun, params, runtime_fun, args_list, opts)
+  rescue
+    e in [UndefinedFunctionError] ->
+      raise_missing_callback(e, :__jit__, 5, __STACKTRACE__)
+  end
+
+  defp raise_missing_callback(exception, name, arity, stacktrace) do
+    case exception do
+      %UndefinedFunctionError{module: compiler, function: ^name, arity: ^arity} ->
+        raise ArgumentError,
```

Comment:
@josevalim I used ArgumentError because this is directly caused by an invalid option being passed. If you have another suggestion for the exception type, I'm open to it!